src/braket/circuits/circuit.py

```diff
@@ -1043,25 +1050,110 @@ def diagram(self, circuit_diagram_class=AsciiCircuitDiagram) -> str:
+            properties don't correspond to the `ir_type`.
         """
-        ir_instructions = [instr.to_ir() for instr in self.instructions]
-        ir_results = [result_type.to_ir() for result_type in self.result_types]
+        if ir_type == IRType.JAQCD:
+            return self._to_jaqcd()
+        elif ir_type == IRType.OPENQASM:
+            if serialization_properties and not isinstance(
+                serialization_properties, OpenQASMSerializationProperties
+            ):
+                raise ValueError(
+                    "serialization_properties must be of type OpenQASMSerializationProperties "
+                    "for IRType.OPENQASM."
+                )
+            return self._to_openqasm(serialization_properties or OpenQASMSerializationProperties())
+        else:
+            raise ValueError(f"Supplied ir_type {ir_type} is not supported.")
+
+    def _to_jaqcd(self) -> JaqcdProgram:
+        jaqcd_ir_type = IRType.JAQCD
+        ir_instructions = [instr.to_ir(ir_type=jaqcd_ir_type) for instr in self.instructions]
+        ir_results = [result_type.to_ir(ir_type=jaqcd_ir_type) for result_type in self.result_types]
         ir_basis_rotation_instructions = [
-            instr.to_ir() for instr in self.basis_rotation_instructions
+            instr.to_ir(ir_type=jaqcd_ir_type) for instr in self.basis_rotation_instructions
         ]
-        return Program.construct(
+        return JaqcdProgram.construct(
             instructions=ir_instructions,
             results=ir_results,
             basis_rotation_instructions=ir_basis_rotation_instructions,
         )
 
+    def _to_openqasm(
+        self, serialization_properties: OpenQASMSerializationProperties
+    ) -> OpenQasmProgram:
+        ir_instructions, qubit_reference_format = self._create_openqasm_header(
+            serialization_properties
+        )
+        serialization_properties.qubit_reference_format = qubit_reference_format
+        openqasm_ir_type = IRType.OPENQASM
+        ir_instructions.extend(
+            [
+                instruction.to_ir(
+                    ir_type=openqasm_ir_type, serialization_properties=serialization_properties
+                )
+                for instruction in self.instructions
+            ]
+        )
+
+        if self.result_types:
+            ir_instructions.extend(
+                [
+                    result_type.to_ir(
+                        ir_type=openqasm_ir_type, serialization_properties=serialization_properties
+                    )
+                    for result_type in self.result_types
+                ]
+            )
+        else:
+            for idx, qubit in enumerate(self.qubits):
+                qubit_target = qubit_reference_format.format(int(qubit))
+                ir_instructions.append(f"b[{idx}] = measure {qubit_target};")
+
+        return OpenQasmProgram.construct(source="\n".join(ir_instructions))
+
+    def _create_openqasm_header(
```

Comment:
I would move this to a create_header() function in the properties, I think.

Reply:
This method has logic related to the circuit serializing itself, so I think it might be better suited here. Since this is private we can move it to the properties later if needed.